server/lib/chat/chat-service.test.ts

```diff
@@ -1314,6 +1382,47 @@ describe('chat/chat-service', () => {
         expectItWorks(result)
       })
     })
+
+    test('should include deleted channels if there are any in chat message', async () => {
+      await joinUserToChannel(
+        user1,
+        testChannel,
+        user1TestChannelEntry,
+        joinUser1TestChannelMessage,
+      )
+
+      const DELETED_ID = makeSbChannelId(999)
+      const message: FakeDbTextChannelMessage = {
+        msgId: 'MESSAGE_5_ID',
+        userId: user1.id,
+        userName: user1.name,
+        channelId: testChannel.id,
+        sent: new Date('2023-03-11T00:00:00.000Z'),
+        data: {
+          type: ServerChatMessageType.TextMessage,
+          text: `Join <#${shieldBatteryChannel.id}> <#${testChannel.id}> and <#${DELETED_ID}>.`,
+          channelMentions: [shieldBatteryChannel.id, testChannel.id, DELETED_ID],
+        },
+      }
+
+      asMockedFunction(getMessagesForChannel).mockResolvedValueOnce([message])
```

Comment:
I'm sure you do this elsewhere in this test, but `mockResolvedValueOnce` is extremely brittle, you're writing tests that depend VERY HEAVILY on the ordering and amount of calls to a method, i.e. they are directly dependent on the code they are testing not changing. I think you'd be better off building a "fake DB" of users for the `findUsersById` call (I don't know how many unique users you have in your tests here but I imagine it's like less than 5, this would be a very easy mock to write.

The other stuff here, I'm pretty sure you could just use `mockResolvedValue` for

Reply:
This is actually the one place in this test file where I used `mockResolvedValueOnce` by mistake :d

I just checked all my other usages of it, and most of them (all but one) are used to mock the function that gets called twice in a same method, e.g. mocking `getUserChannelEntryForUser` (this is the majority of its usages) in multiple methods that call it like this:
```ts
const [channelInfo, userChannelEntry, targetChannelEntry] = await Promise.all([
  getChannelInfo(channelId),
  getUserChannelEntryForUser(userId, channelId),
  getUserChannelEntryForUser(targetId, channelId),
])
```

Similarly for `findUsersById`:
```ts
const [users, userMentions, channelMentions] = await Promise.all([
  findUsersById(Array.from(userIds)),
  findUsersById(Array.from(userMentionIds)),
  this.getChannelInfos(Array.from(channelMentionIds), userId),
])
```

I was hesitant to use `mockResolvedValueOnce` myself because I didn't like how it relied on stack call order, but I thought it would be fine here since these calls are right next to each other, and Jest documentation says this is the easiest way to mock a (different) return value of a function that was called multiple times. Also, I think it shows the intent quite nicely. For example, when I want to test that the target user is a channel moderator, I just use it like this:
```ts
asMockedFunction(getUserChannelEntryForUser)
  .mockResolvedValueOnce(user1TestChannelEntry)
  .mockResolvedValueOnce({
    ...user2TestChannelEntry,
    channelPermissions: { ...channelPermissions, editPermissions: true },
})
```
I'm not opposed to creating a fake DB(s?) to replace all of the `mockResolvedOnce` usage, but I feel like it would take me some time (not because it's necessarily complicated, but because I don't know how to do it atm :d).

Reply:
I think in those it would generally be better to just write a simple implementation. When I was talking about a fake, to be clear, I was talking about one that would be read-only, which is significantly easier to implement correctly than the other way (but you'd only be using this for data the code depends on that it never actually changes).

For the thing I mentioned in particular, it is absurdly easy to do, and I've already done it in other tests (e.g. see the game loader branch), it's just not easy to share an implementation between tests due to Jest's rules about mocks not accessing external data.

For the stuff you do expect to change (e.g. the last example you have here), I think the better thing to do is just write a simple implementation in the test, where you've written the `mockResolvedValueOnce` now. e.g.

```ts
asMockedFunction(getChannelEntryForUser).mockImplementation(async (channelId, userId) => {
  if (userId === USER_1) {
    return user1TestChannelEntry
  } else if (userId == USER_2) {
    return user2TextChannelEntry
  } else {
    throw new Error('Unexpected user lookup')
  }
})
```